game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/AaCasualtySelector.java

```diff
@@ -114,265 +112,209 @@ private static CasualtyDetails getLowLuckAaCasualties(
       unitHp.computeIfPresent(
           unit,
           (unitKey, hp) -> {
             if (hp > 1) {
-              finalCasualtyDetails.addToDamaged(unit);
+              casualtyDetails.addToDamaged(unit);
             } else {
-              finalCasualtyDetails.addToKilled(unit);
+              casualtyDetails.addToKilled(unit);
             }
             return hp - 1;
           });
     }
-    return finalCasualtyDetails;
+    return casualtyDetails;
   }
 
-  /**
-   * http://triplea.sourceforge.net/mywiki/Forum#nabble-td4658925%7Ca4658925 returns two lists, the
-   * first list is the air units that can be evenly divided into groups of 3 or 6 (depending on
-   * radar) the second list is all the air units that do not fit in the first list
-   */
-  private static Tuple<List<List<Unit>>, List<Unit>> categorizeLowLuckAirUnits(
-      final Collection<Unit> units, final int groupSize) {
-    final Collection<UnitCategory> categorizedAir =
-        UnitSeparator.categorize(units, null, false, true);
-    final List<List<Unit>> groupsOfSize = new ArrayList<>();
-    final List<Unit> toRoll = new ArrayList<>();
-    for (final UnitCategory uc : categorizedAir) {
-      final int remainder = uc.getUnits().size() % groupSize;
-      final int splitPosition = uc.getUnits().size() - remainder;
-      final List<Unit> group = new ArrayList<>(uc.getUnits().subList(0, splitPosition));
-      if (!group.isEmpty()) {
-        for (int i = 0; i < splitPosition; i += groupSize) {
-          final List<Unit> miniGroup = new ArrayList<>(uc.getUnits().subList(i, i + groupSize));
-          if (!miniGroup.isEmpty()) {
-            groupsOfSize.add(miniGroup);
-          }
-        }
-      }
-      toRoll.addAll(uc.getUnits().subList(splitPosition, uc.getUnits().size()));
+  private static CasualtyDetails calculateRolledAaCasualties(
+      final List<Unit> availableTargets,
+      final AaPowerStrengthAndRolls unitPowerAndRollsMap,
+      final DiceRoll dice,
+      final IDelegateBridge bridge) {
+
+    final int hits = dice.getHits();
+    final Collection<Unit> hitTargets;
+    if (unitPowerAndRollsMap.calculateTotalRolls() == availableTargets.size()
+        && hits < availableTargets.size()) {
+      // there is a roll for every target but not enough hits to kill all of the targets
+      // so no need to get a random set of units since all units will either have a hit
+      // or miss roll
+      hitTargets = findRolledTargets(availableTargets, dice);
+    } else if (hits < availableTargets.size()) {
+      // there isn't a roll for every target so need to randomly pick the target for each hit
+      hitTargets = findRandomTargets(availableTargets, bridge, hits);
+    } else {
+      // all targets were hit so add them all
+      hitTargets = availableTargets;
     }
-    return Tuple.of(groupsOfSize, toRoll);
+
+    return buildCasualtyDetails(availableTargets, hitTargets);
```

Comment:
Perhaps a style preference on my part, to another extent I see we are doing the 'single-return' point design as well. I think some this method block would be relatively clear returning when we can. For example, the last 'else' statement would avoid some extra thought to realize that `availableTargets' is equal to 'hitTargets'.

Personally I've come to find the single-return point design is only really good for longer methods, ones that are complicated and return for different reasons at different places. For short methods, something like the below is just fine IMO:
```
if
   return
else if
   return
else
   return
```

The above example I'd say is more clear than the single return variable, ie:
```
var returnVal;
if
   returnVal = a
else if
   returnVal =b
else
   returnVal =c

return returnVal
```